zaza/model.py

```diff
@@ -206,6 +204,93 @@ async def run_in_model(model_name):
         await model.disconnect()
 
 
+async def block_until_auto_reconnect_model(*conditions,
+                                           model=None,
+                                           aconditions=None,
+                                           timeout=None,
+                                           wait_period=0.5,
+                                           loop=None):
+    """Async block on the model until conditions met.
+
+    This function doesn't use model.block_until() which unfortunately raises
+    websockets.exceptions.ConnectionClosed if the connections gets closed,
+    which seems to happen quite frequently.  This funtion blocks until the
+    conditions are met or a timeout occurs, and reconnects the model if it
+    becomes disconnected.
+
+    Note that conditions are just passed as an unamed list in the function call
+    to make it work more like the more simple 'block_until' function.
+
+    :param model: the model to use OR None, in which case the model_name is
+        also optionally used.
+    :type model: :class:'juju.Model()'
+    :param conditions: a list of callables that need to evaluate to True.
+    :type conditions: [List[Callable[[:class:'juju.Model()'], bool]]]
+    :param aconditions: an optional list of async callables that need to
+        evaluate to True.
+    :type aconditions:
+        Optional[List[AsyncCallable[[:class:'juju.Model()'], bool]]]
+    :param timeout: the timeout to wait for the block on.
+    :type timeout: float
+    :param wait_period: The time to sleep between checking the conditions.
+    :type wait_period: float
+    :param loop: The event loop to use
+    :type loop: An event loop
+    :raises: TimeoutError if the conditions never match (assuming timeout is
+        not None).
+    """
+    assert model is not None, ("model can't be None in "
+                               "block_until_auto_reconnect_model()")
+    model_name = model.info.name
+
+    def _disconnected():
+        return not (model.is_connected() and model.connection().is_open)
+
+    aconditions = aconditions or []
+
+    def _done():
+        return all(c() for c in conditions)
+
+    async def _adone():
+        evaluated = []
+        # note Python 3.5 doesn't support async comprehensions; do it the old
+        # fashioned way.
+        for c in aconditions:
+            evaluated.append(await c())
+            if _disconnected():
+                return True
```

Comment:
Shouldn't we `return False` instead? I know it's supposed to never happen, but I'd say if we're disconnected and can't determine if the conditions are fulfilled or not, I'd rather assume they are **not fulfilled**. Or am I getting this wrong?

I know this will be caught on line 285 anyway but it feels cleaner to me to return False here.

Reply:
You're right, it doesn't actually matter; it was just to short-circuit the checks and get back to trying to re-connect.  However, False makes more sense.